### PR TITLE
feat: Use SecretProvider to get Config/Registry access tokens

### PR DESCRIPTION
**Pull request by lenny-intel**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/go-mod-bootstrap/blob/master/.github/Contributing.md.

## What is the current behavior?
SecretProvider currently create via bootstrap handler long after Config and Registry clients are created.

## What is the new behavior?
SecretProvider now created during config processing and used to get access tokens when creating the Config and registry clients

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## New Imports

- [ ] Yes
- [x] No

## Specific Instructions
Are there any specific instructions or things that should be known prior to reviewing?

## Other information

### CODE REVIEW

File: bootstrap/config/config.go [hunk 1]
```diff
@@ -130,15 +141,19 @@ func (cp *Processor) Process(serviceKey string, configStem string, serviceConfig
...
+			accessToken, err = secretProvider.GetAccessToken(configProviderInfo.serviceConfig.Type, serviceKey)
+			if err != nil {
+				return fmt.Errorf(
+					"failed to get Configuration Provider (%s) access token: %s",
+					configProviderInfo.serviceConfig.Type,
+					err.Error())
 			}
+
+			cp.lc.Infof("Using Config Provider access token of length %d", len(accessToken))
 		}
```
Comment (jim-wang-intel) [inline bootstrap/config/config.go:157]: request to add cp.lc.Info for else branch like "not configured to use secret provider"

File: bootstrap/config/config.go [hunk 2]
```diff
@@ -288,25 +303,25 @@ func (cp *Processor) ListenForCustomConfigChanges(
...
 			case ex := <-errorStream:
-				cp.Logger.Error(ex.Error())
+				cp.lc.Error(ex.Error())
 
 			case raw := <-updateStream:
 				//if ok := configToWatch.UpdateWritableFromRaw(raw); !ok {
-				//	cp.Logger.Error("unable to update custom writable configuration from Configuration Provider")
+				//	cp.lc.Error("unable to update custom writable configuration from Configuration Provider")
 				//	continue
 				//}
```
Comment (jim-wang-intel) [inline bootstrap/config/config.go:316]: remove the dead code?

File: bootstrap/config/config.go [hunk 3]
```diff
@@ -479,5 +494,5 @@ func (cp *Processor) listenForChanges(serviceConfig interfaces.Configuration, co
 
 // logConfigInfo logs the config info message with number over overrides that occurred.
 func (cp *Processor) logConfigInfo(message string, overrideCount int) {
-	cp.Logger.Info(fmt.Sprintf("%s (%d envVars overrides applied)", message, overrideCount))
+	cp.lc.Info(fmt.Sprintf("%s (%d envVars overrides applied)", message, overrideCount))
```
Comment (jim-wang-intel) [inline bootstrap/config/config.go:497]: Infof?

File: bootstrap/interfaces/secret.go [hunk 4]
```diff
@@ -16,4 +16,6 @@ type SecretProvider interface {
 
 	// SecretsLastUpdated returns the last time secrets were updated
 	SecretsLastUpdated() time.Time
+
+	GetAccessToken(tokenType string, serviceKey string) (string, error)
```
Comment (jim-wang-intel) [inline bootstrap/interfaces/secret.go:20]: please add doc/comment for this interface

**Review by jim-wang-intel — COMMENTED** [review]
suggested minor tweaks

File: bootstrap/config/config.go (hunk 3 above)
Comment (lenny-intel) [reply bootstrap/config/config.go:497]: Yep, good catch! THX!

File: bootstrap/config/config.go (hunk 2 above)
Comment (lenny-intel) [reply bootstrap/config/config.go:316]: Removed

File: bootstrap/config/config.go (hunk 1 above)
Comment (lenny-intel) [reply bootstrap/config/config.go:157]: Added

File: bootstrap/interfaces/secret.go (hunk 4 above)
Comment (lenny-intel) [reply bootstrap/interfaces/secret.go:20]: Yep, missed that on.. good catch

**Review by jim-wang-intel — APPROVED** [review]
LGTM

File: bootstrap/config/config.go [hunk 5]
```diff
@@ -130,15 +141,21 @@ func (cp *Processor) Process(serviceKey string, configStem string, serviceConfig
...
+		if secretProvider != nil {
+			accessToken, err = secretProvider.GetAccessToken(configProviderInfo.serviceConfig.Type, serviceKey)
+			if err != nil {
+				return fmt.Errorf(
+					"failed to get Configuration Provider (%s) access token: %s",
+					configProviderInfo.serviceConfig.Type,
+					err.Error())
 			}
+
+			cp.lc.Infof("Using Config Provider access token of length %d", len(accessToken))
```
Comment (bnevis-i) [inline bootstrap/config/config.go:156]: Out of curiosity, now what we have access token, how is it being cached?

**Review by bnevis-i — APPROVED** [review]

File: bootstrap/config/config.go (hunk 5 above)
Comment (lenny-intel) [reply bootstrap/config/config.go:156]: Not caching it to reuse between Config and Registry client. They each get their own token. 

The clients are configured with the access tokens, so technically the tokens are cached in each of the clients config.

Comment (bnevis-i) [reply bootstrap/config/config.go:156]: ok.  So each edgeX service nomally gets two consul tokens, one for the registry client, one for the config client?   No response necessary unless I got it wrong.

Comment (lenny-intel) [reply bootstrap/config/config.go:156]: Correct

**Pull request merged**